torch/fx/passes/_tensorify_python_scalars.py

```diff
@@ -0,0 +1,262 @@
+from __future__ import annotations
+
+import logging
+from typing import List, Union
+
+import torch
+import torch.fx as fx
+from torch._prims_common import get_computation_dtype
+from torch._subclasses import fake_tensor  # noqa: TCH001
+from torch.fx._utils import lazy_format_graph_code
+from torch.fx.experimental.symbolic_shapes import ShapeEnv  # noqa: TCH001
+from torch.fx.graph_module import GraphModule  # noqa: TCH001
+
+# TODO: refactor
+from torch.fx.passes.runtime_assert import _get_sym_val
+from torch.fx.proxy import MetaProxy
+from torch.utils._sympy.reference import TensorReferenceAnalysis
+
+
+__all__: List[str] = []
+
+log = logging.getLogger(__name__)
+graph_code_log = torch._logging.getArtifactLogger(__name__, "graph_code")
+
+# The general shape of this transformation is to look for Tensor operations
+# that take a backed SymFloat as an argument, and then redo them as tensor
+# compute (with ints and tensors as inputs). For example, add(Tensor, Scalar)
+# can be translated into add(Tensor, Tensor). Because Dynamo has already
+# arranged for floats to be Tensor inputs to the graph, for typical float
+# compute you can entirely translate the Python float operations into Tensor
+# operations with only Tensor inputs.
+#
+# This pass is also responsible for doing CSE on the fly as we do this, since
+# you don't want to keep recomputing the same quantity over and over again if
+# it's used multiple times.
+#
+# This pass runs on the JOINT graph produced by AOT Autograd, prior to
+# partitioning, because we want to be able to make changes that affect
+# our partitioning decisions (in particular, we want to avoid having to
+# be able to save floats across the partition, and passes that change what
+# device compute happen on need to happen before partitioning, but after this
+# pass). Note that some transformations have to happen before this in Dynamo,
```

Comment:
just for my understanding, which transformations do we need to do earlier in dynamo to avoid specialization? (vs. the the transformation here that we are ok with delaying until we get the joint graph)

Reply:
Dynamo will convert float inputs into 0d CPU tensor inputs, which then immediately have item() calls to transform them into regular float for downstream use: https://github.com/pytorch/pytorch/pull/125325

Reply:
ah right :) thanks